pkg/client/apiutil/restmapper.go

```diff
@@ -162,25 +167,34 @@ func (m *mapper) addKnownGroupAndReload(groupName string, versions ...string) er
 	// If no specific versions are set by user, we will scan all available ones for the API group.
 	// This operation requires 2 requests: /api and /apis, but only once. For all subsequent calls
 	// this data will be taken from cache.
-	if len(versions) == 0 {
-		apiGroup, err := m.findAPIGroupByName(groupName)
+	//
+	// Calling this also tells us if the server supports aggregated discovery, which allows
+	// loading everything with two api calls which we assume is overall cheaper, hence we always
+	// do this once optimistically.
+	if len(versions) == 0 || !m.initialDiscoveryDone.Load() {
+		apiGroup, didAggregatedDiscovery, err := m.findAPIGroupByNameAndMaybeAggregatedDiscovery(groupName)
 		if err != nil {
 			return err
 		}
-		if apiGroup != nil {
+		if apiGroup != nil && len(versions) == 0 {
 			for _, version := range apiGroup.Versions {
 				versions = append(versions, version.Version)
 			}
 		}
-	}
-
-	m.mu.Lock()
-	defer m.mu.Unlock()
 
-	// Create or fetch group resources from cache.
-	groupResources := &restmapper.APIGroupResources{
-		Group:              metav1.APIGroup{Name: groupName},
-		VersionedResources: make(map[string][]metav1.APIResource),
+		// No need to do anything further if aggregatedDiscovery is supported and we did a lookup
+		if didAggregatedDiscovery {
+			failedGroups := make(map[schema.GroupVersion]error)
+			for _, version := range versions {
+				if m.knownGroups[groupName] == nil || m.knownGroups[groupName].VersionedResources[version] == nil {
+					failedGroups[schema.GroupVersion{Group: groupName, Version: version}] = &meta.NoResourceMatchError{}
+				}
+			}
```

Comment:
(I intentionally put this comment on a different line to start a new conversation)

WDYT about setting `PartialResource` in `NoResourceMatchError`. I think if we set a GVR with group and version we can get slightly better error messages

Reply:
Added